core/src/main/java/fj/data/List.java

```diff
@@ -563,7 +562,57 @@ public final void foreachDoEffect(final Effect1<A> f) {
     return bind(c);
   }
 
-  /**
+    /**
+     * Traverse through the List with a might-fail function.
+     *
+     * @param f The function that might fail.
+     * @return  none if applying f fails to any element of the list or f mapped list in some .
+     */
+    public <B> Option<List<B>> traverseOption(final F<A, Option<B>> f) {
+        F2<A, Option<List<B>>, Option<List<B>>> f2 = (a, as) -> map2Option(List::cons, f.f(a), as);
+        return foldRight(f2, Option.some(List.<B>nil()));
+    }
+
+    private static <A, B, C> Option<List<C>> map2Option(F2<A, List<B>, List<C>> f, final Option<A> opt1, final Option<List<B>> opt2) {
+        return opt1.bind(o1 -> opt2.map(o2 -> f.f(o1, o2)));
+    }
+
+    /**
+     * Traverse through the List with a function that might fail with reporting an error.
+     *
+     * @param f The function that produces Either value.
+     * @return  error in left if applying f fails to any element of the list or f mapped list in right.
+     */
+    public <B, E> Either<E, List<B>> traverseEither(final F<A, Either<E, B>> f) {
+        F2<A, Either<E, List<B>>, Either<E, List<B>>> f2 = (A a, Either<E, List<B>> as) -> List.<B, B, B, E>map2Either(List::cons, f.f(a), as);
+        return foldRight(f2, Either.<E, List<B>>right(List.<B>nil()));
+    }
+
+    private static <A, B, C, E> Either<E, List<C>> map2Either(F2<A, List<B>, List<C>> f, final Either<E, A> either1, final Either<E, List<B>> either2) {
+        return either1.right().bind(o1 -> either2.right().map(o2 -> f.f(o1, o2)));
+    }
+
+    public <B> Stream<List<B>> traverseStream(final F<A, Stream<B>> f) {
+        F2<A, Stream<List<B>>, Stream<List<B>>> f2 = (A a, Stream<List<B>> as) -> List.<B, B, B>map2Stream(List::cons, f.f(a), as);
+        return foldRight(f2, Stream.<List<B>>nil());
+    }
+
+    private static <A, B, C> Stream<List<C>> map2Stream(F2<A, List<B>, List<C>> f, final Stream<A> stream1, final Stream<List<B>> stream2) {
+        return stream1.bind(o1 -> stream2.map(o2 -> f.f(o1, o2)));
+    }
+
+    public <B> P1<List<B>> traverseP1(final F<A, P1<B>> f){
+        F2<A, P1<List<B>>, P1<List<B>>> f2 = (A a, P1<List<B>> ps) -> f.f(a).bind(b -> ps.map(bs -> bs.cons(b)));
+        return foldRight(f2, P.p(List.<B>nil()));
+    }
+
+    public <B> IO<List<B>> traverseIO(F<A, IO<B>> f) {
+        F2<A, IO<List<B>>, IO<List<B>>> f2 = (a, acc) ->
+                IOFunctions.bind(acc, (bs) -> IOFunctions.map(f.f(a), b -> bs.cons(b)));
+        return this.foldRight(f2, IOFunctions.unit(List.<B>nil()));
+    }
```

Comment:
Now I have looked further down, this looks the same as my way above, but all on the one line to avoid having to specify the types.